src/counter.rs

```diff
@@ -0,0 +1,69 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::{Duration, UNIX_EPOCH, SystemTime};
+
+pub struct Counter {
+    pub name: &'static str,
+    pub counts: AtomicUsize,
+    pub nanos: AtomicUsize,
+    pub times: AtomicUsize,
+    pub lograte: usize,
+}
+
+macro_rules! STATIC_COUNTER {
+    ($name:expr, $lograte:expr) => {
+        Counter {
+            name: $name,
+            counts: AtomicUsize::new(0),
+            nanos: AtomicUsize::new(0),
+            times: AtomicUsize::new(0),
+            lograte: $lograte,
+        }
+    };
+}
+
+macro_rules! STATIC_COUNTER_INC {
+    ($name:expr, $count:expr, $start:expr) => {
+        unsafe { $name.inc($count, $start.elapsed()) };
+    };
+}
+
+impl Counter {
+    pub fn inc(&mut self, events: usize, dur: Duration) {
+        let total = dur.as_secs() * 1_000_000_000 + dur.subsec_nanos() as u64;
+        let counts = self.counts.fetch_add(events, Ordering::Relaxed);
+        let nanos = self.nanos.fetch_add(total as usize, Ordering::Relaxed);
+        let times = self.times.fetch_add(1, Ordering::Relaxed);
+        if times % self.lograte == 0 && times > 0 {
+            let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
```

Comment:
Can you use `Instant` here, so that it's consistent with other uses of `now()` within the codebase?

Reply:
what would the instant be referencing as epoch?  each counter needs to print its value against a global point of reference of time.

Reply:
can you use timing::timestamp() does what this does